File: snapshots/skipstone/kickbasecore/KickbaseCore/destination/skipstone/KickbaseCore/src/main/kotlin/kickbase/core/SearchUtils.kt
package kickbase.core

import skip.lib.*
import skip.lib.Array

import skip.foundation.*

// Helper to extract content safely without exposing Swift.String.Index to Kotlin/Skip

fun String.substringAfter(marker: String): String? {
    // Safer implementation effectively using split/components which transpiles well
    val parts = this.components(separatedBy = marker)
    if (parts.count <= 1) {
        return null
    }
    // Join the rest in case the marker appears multiple times
    return parts.dropFirst().joined(separator = marker)
}

fun String.substringBefore(marker: String): String? {
    val parts = this.components(separatedBy = marker)
    if (parts.count <= 1) {
        return null
    }
    return parts.first
}

fun String.substringBetween(startMarker: String, endMarker: String): String? {
    val after_0 = this.substringAfter(startMarker)
    if (after_0 == null) {
        return null
    }
    return after_0.substringBefore(endMarker)
}

fun String.splitBy(separator: String): Array<String> = this.components(separatedBy = separator)

fun Substring.substringAfter(marker: String): String? = String(this).substringAfter(marker)

fun Substring.substringBefore(marker: String): String? = String(this).substringBefore(marker)

fun Substring.substringBetween(startMarker: String, endMarker: String): String? = String(this).substringBetween(startMarker, endMarker)

fun Substring.splitBy(separator: String): Array<String> = String(this).components(separatedBy = separator)

File: KickbaseCore/Sources/KickbaseCore/Utilities/SearchUtils.swift
import Foundation

extension String {
    // Helper to extract content safely without exposing Swift.String.Index to Kotlin/Skip

    public func substringAfter(_ marker: String) -> String? {
        // Safer implementation effectively using split/components which transpiles well
        let parts = self.components(separatedBy: marker)
        guard parts.count > 1 else { return nil }
        // Join the rest in case the marker appears multiple times
        return parts.dropFirst().joined(separator: marker)
    }

    public func substringBefore(_ marker: String) -> String? {
        let parts = self.components(separatedBy: marker)
        guard parts.count > 1 else { return nil }
        return parts.first
    }

    public func substringBetween(_ startMarker: String, _ endMarker: String) -> String? {
        guard let after = self.substringAfter(startMarker) else { return nil }
        return after.substringBefore(endMarker)
    }

    public func splitBy(_ separator: String) -> [String] {
        return self.components(separatedBy: separator)
    }

    /// Split a string by any of the provided characters (e.g., " _-") in a stable, explicit way
    public func splitByCharacters(_ separators: String) -> [String] {
        var out: [String] = []
        var current = ""
        for ch in self {
            if separators.contains(ch) {
                if !current.isEmpty {
                    out.append(current)
                    current = ""
                }
            } else {
                current.append(ch)
            }
        }
        if !current.isEmpty { out.append(current) }
        return out
    }
}

extension Substring {
    public func substringAfter(_ marker: String) -> String? {
        return String(self).substringAfter(marker)
    }

    public func substringBefore(_ marker: String) -> String? {
        return String(self).substringBefore(marker)
    }

    public func substringBetween(_ startMarker: String, _ endMarker: String) -> String? {
        return String(self).substringBetween(startMarker, endMarker)
    }

    public func splitBy(_ separator: String) -> [String] {
        return String(self).components(separatedBy: separator)
    }

    public func splitByCharacters(_ separators: String) -> [String] {
        return String(self).splitByCharacters(separators)
    }
}
